[do not merge] Initial implementation of the @_specialize attribute supporting layout constraints and partial specializations.

swiftix opened the pull request:
One can use the `@_specialize` attribute to force creation of specializations of a given function for a given set of generic parameters.

The new `@_specialize` attribute has a syntax like this:

```swift
@_specialize(exported: false, kind: full, where K == Int, V == Int)
@_specialize(exported: false, kind: partial, where K: _Trivial64)
func dictFunction<K, V>(dict: Dictionary<K, V>) {
}
```

If `exported` is set, the corresponding specialization would have a public visibility and can be used by clients.
If `exported` is omitted, it's value is assumed to be `false`.

If `kind` is `full` it means a full specialization.
If `kind` is `partial` it means a partial specialization.
If `kind` is omitted, its value is assumed to be `full`.

One can use in the `where` clause of the `@_specialize` attribute the newly introduced layout constraints like e.g. _Trivial, _Trivial8, _Trivial16, _Trivial32, _Trivial64 or _RefCountedObject. 

These layout attributes have the following semantics:
- `_TrivialN` means an value of a trivial type, which is N bits long. Trivial means that it can be copied by means of a bitwise copy and does not contain any references. If `N` is not specified then object can be of any size, but should be still trivial.
- `_RefCountedObject ` means a reference counted object, e.g. it can be a reference to an instance of a class.

swiftix committed:
Initial version of the partial specialization support.

It is currently hidden behind a flag: -Xllvm -enable-partial-specialization

swiftix committed:
Initial implementation of the @_specialize attribute supporting layout constraints and partial specializations.

One can use the `@_specialize` attribute to force creation of specializations of a given function for a given set of generic parameters.

The new `@_specialize` attribute has a syntax like this:

```swift
@_specialize(exported: false, kind: full, where K == Int, V == Int)
@_specialize(exported: false, kind: partial, where K: _Trivial64)
func dictFunction<K, V>(dict: Dictionary<K, V>) {
...
}
```

If `exported` is set, the corresponding specialization would have a public visibility and can be used by clients.
If `exported` is omitted, it's value is assumed to be `false`.

If `kind` is `full` it means a full specialization.
If `kind` is `partial` it means a partial specialization.
If `kind` is omitted, its value is assumed to be `full`.

One can use in the `where` clause of the `@_specialize` attribute the newly introduced layout constraints like e.g. _Trivial, _Trivial8, _Trivial16, _Trivial32, _Trivial64 or _RefCountedObject.

These layout constraints have the following semantics:
- `_TrivialN` means an value of a trivial type, which is N bits long. Trivial means that it can be copied by means of a bitwise copy and does not contain any references. If `N` is not specified then object can be of any size, but should be still trivial.
- `_RefCountedObject ` means a reference counted object, e.g. it can be a reference to an instance of a class.

swiftix commented:
The first commit is simply cherry-picked #5463, because it is required as a basis. The actual feature implementation is in the second commit.

swiftix commented:
@DougGregor @slavapestov Could you review this PR? In particular, I'd like to hear if the way how I introduced the new layout constraints is OK or if it should be done differently.

slavapestov commented on lib/AST/Attr.cpp:
```diff
@@ -440,9 +441,39 @@ bool DeclAttribute::printImpl(ASTPrinter &Printer, const PrintOptions &Options)
+      auto *SM = FnDecl->getModuleContext();
+      assert(GenericEnv);
+      GetInterfaceType = [=](Type Ty) -> Type {
+        return GenericEnv->mapTypeOutOfContext(
```
There's a getSugaredType() method you can use

swiftix replied:
@slavapestov getSugaredType takes only `GenericTypeParamType` as its parameters. It is impossible to call it with a Type (which may be something containing GenericTypeParamType). So, one would need a `type.walk` here. It is not clear to me if it would better that this MapInto/MapOutOf Context trick.

slavapestov replied:
Mapping a type into context loses information (two type parameters that are related via a same type constraints map to the same archetype) so adding a new version of getSugaredType that calls Type::subst() with a lambda invoking the GenericTypeParamType version would be best.

slavapestov commented on lib/SIL/SILFunctionType.cpp:
```diff
@@ -2231,8 +2235,28 @@ namespace {
+        // the lowering should happen in the target generic context.
+
+        // Temporarily replace the generic context.
+        TheSILModule.Types.popGenericContext(genericSig);
```
I'm getting rid of this push/pop nonsense soon :-)

swiftix replied:
This is a workaround for rdar://29711782. There is a bug in substGenericArgs. If you remove this nonsense, but fix the issue, I'm fine with it ;-)

slavapestov replied:
I'll look at the bug, thanks. I bet it affects more than the optimizer.

slavapestov commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -2328,7 +2328,34 @@ void SILDebugScope::dump(SourceManager &SM, llvm::raw_ostream &OS,
+               [&](Requirement req) {
+                 // Use GenericEnvironment to produce user-friendly
+                 // names instead of something like t_0_0.
+                 auto FirstTy = GenericEnv->mapTypeOutOfContext(
```
Same comment as above

slavapestov commented on lib/SIL/SILVerifier.cpp:
```diff
@@ -751,7 +751,8 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
     if (subs.empty()) {
-      require(!fnTy->isPolymorphic(),
+      require(!fnTy->isPolymorphic() ||
+                  fnTy->getGenericSignature()->areAllParamsConcrete(),
```
When I added concrete extensions my intent was that we would never build a SILFunctionType with a fully concrete generic signature. If this invariant is broken do you mind fixing it at the source?

swiftix replied:
@slavapestov I'm not sure what you suggest here by "fixing it at the source". Do you suggest to use a non-polymorphic  SILFunctionType for such full specializations? Or do you suggest to fix the isPolymorphic test, so that it does not treat a type with a fully generic signature as polymorphic?

slavapestov replied:
The former. If all generic parameters were made concrete in a signature, just drop the signature when the SILFunctionType is constructed. We already do this for methods defined in concrete extensions (e.g. extension Array where Element == Int) so I'd greatly prefer to also do that for full specializations instead of adding a new possible state for SILFunctionType.

slavapestov commented on stdlib/public/core/TypeLayoutConstraints.swift:
```diff
@@ -0,0 +1,40 @@
+// These protocols can be used e.g. in @_specialize attribute.
+
+// Type should be a trivial type. It's size should be 8 bits.
+public protocol _Trivial8 {
```
Are these still part of the design?

slavapestov replied:
Ping? It is not clear to me how these protocols are used. In principle we should need to represent these layout constraints with protocols since it will cause us to pass unnecessary witness tables, unless we do more work in IRGen to special case them (which is something I'd like to avoid, special behavior of AnyObject is enough of a hassle already)

slavapestov replied:
I see your other comment now. Yeah, we should find a way to represent these constraints as their own thing and not protocols. You already added code to ArchetypeBuilder to handle them -- why are the protocols needed?

swiftix replied:
@slavapestov If we do not represent these constraints internally as protocols, what would you use to represent them? Are you saying we should not represent them as types at all, but e.g. as special attributes? Or do you suggest to represent them as special (internal) types? Can you elaborate?

swiftix replied:
@slavapestov I had a discussion with @DougGregor about this. I'll represent those layout constraints as non-types and adjust `Requirement` and `RequirementRepr` accordingly.

slavapestov commented on include/swift/AST/LayoutConstraint.h:
```diff
@@ -0,0 +1,107 @@
+enum class LayoutConstraintKind {
+  // It is not a known layout constraint.
+  UnknownLayout,
+  Trivial8 = 1,
```
Can we allow any size instead of offering a fixed set of sizes? Or at least make the attribute syntax look more general than it is.

swiftix replied:
Yes, it should be pretty-straight forward to add this in the future. But since syntax and semantics of these new attributed were discussed and approved on the swift-evolution yet, I decided to go for a simple approach for now. 

But I'd love to have a more general way to express the size.

swiftix replied:
@slavapestov Btw, @rjmccall had the idea to allow for layout constraints like `_TrivialOfSizeAtMostN`, because it can be very handy. I'll see what I can do. But it is an additive feature I could add later.

swiftix replied:
@slavapestov Regarding the more general syntax, I guess something like `_Trivial(N)` would be nice. But currently I expressed these layout constraints as protocols. And we don't have generic protocols yet :-( Otherwise one could just use `_Trivial<N>` instead. To make these constraints parametrized right now, one would need to treat them as something very special, not as protocols. I'm not sure it is a good trade-off. But if we think it is worth it, we could change it.

slavapestov replied:
I don't think this stuff needs to go through swift-evolution, it is best to keep it as an underscored attribute since it is closely tied to the optimizer implementation, it will give us freedom to change it in the future and avoid the inevitable bikeshedding and unreasonable feature requests :-)

slavapestov commented on include/swift/AST/LayoutConstraint.h:
```diff
@@ -0,0 +1,107 @@
+  // It is a layout constraint representing a trivial type of an unknown size.
+  Trivial = 1024,
+  // It is a layout constraint representing a reference counted object.
+  RefCountedObject,
```
It might be worthwhile to distinguish native from unknown refcounting too.

swiftix replied:
@slavapestov Yes, this is a good point!

slavapestov reviewed (commented):
I skimmed this and it looks good. I'll take a more detailed look later. There are some commented out code blocks and #if 0 etc that I assume you will clean up. Other than that I only have a few comments so far.

brentdax commented:
Perhaps I'm missing something, but why do you have the `kind: full/partial` parameter? It seems to me that the same information is available by analyzing the `where` clause—either it constrains all the generic parameters tightly enough that the function is fully specialized, or it leaves them incompletely constrained and it must be partially specialized.

Are there cases where the same `where` clause could be used for full or partial specialization, and the parameter is necessary to distinguish between them? Is it too difficult to check whether a `where` clause would fully constrain the parameters? Or is there some other purpose to this?

slavapestov commented:
@brentdax GenericSignature::areAllParamsConcrete() tests if all parameters have been constrained or not.

swiftix commented:
@brentdax Your observation is correct. If the `kind` parameter would be introduced only for the compiler, then it wouldn't be required, because the compiler could easily derive this information by analyzing the `where`-clause. 

But I introduced this parameter for developers. It is very easy to forget one of the generic parameters in a `where`-clause even if you actually meant a full specialization (e.g. in nested generic contexts, etc). Therefore, I think it is much safer to explicitly call out if one really intends to produce a partial specialization.

brentdax commented:
@swiftix Then I wonder if we might be better off with a syntax like `@_specialize(exported: false, partial, where …)` (with no corresponding `full`, merely the absence of the `partial` keyword), or perhaps even `@_specialize(public, partial, where …)`.

swiftix commented:
@brentdax Yes, this could work. I do not insist on the exact syntax that I implemented so far. I decided to use `kind: partial`, because it is more in line with other attributes and named parameters in Swift. BTW, @jrose-apple was very much in favor of using this more regular syntax.

xwu commented:
Perhaps one could be even terser. Looking to `@available` for inspiration, we could have `*` indicate that other generic parameters are deliberately omitted; for instance:

```swift
@_specialize(where K: _Trivial64, *)
```

A diagnostic can tell the developer to specify all generic parameters or append `*`. Combine that with @brentdax's suggestion for `public` and we can get rid of that unslightly comma before `where` :)

swiftix committed:
Address some of the issues noticed in @slavapestov's review.

jrose-apple commented:
To be clear, I'm in favor of *syntax that looks like a normal attribute,* but I don't think that means all "arguments" have to be labeled. It's also not that important right now; if we want to expose this for general use we'll be putting it through swift-evolution.

swiftix committed:
Minor cosmetic fixes after re-basing.

slavapestov commented on include/swift/AST/ArchetypeBuilder.h:
```diff
@@ -154,17 +148,23 @@ class ArchetypeBuilder {
                                         Type Concrete,
                                         RequirementSource Source);
 
+  /// \brief Add a new same-type requirement specifying that the given potential
```
Do these still need to be public?

slavapestov commented on include/swift/AST/LayoutConstraint.h:
```diff
@@ -0,0 +1,134 @@
+LayoutConstraintInfo getLayoutConstraintInfo(Type Ty);
+
+/// Checks if a given TypeRepr is a layout constraint type.
+bool isLayoutConstraintType(TypeRepr *TyR);
```
Instead of layout constraint types can we just parse and store the layout constraint byte size, etc directly?

slavapestov commented on include/swift/AST/Types.h:
```diff
@@ -3273,8 +3273,11 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
-  bool isPolymorphic() const { return GenericSig != nullptr; }
+
+  bool isPolymorphic() const {
+    return GenericSig != nullptr && !GenericSig->areAllParamsConcrete();
```
Why not just build a SILFunctionType without a GenericSignature in this case?

slavapestov commented on include/swift/Parse/Parser.h:
```diff
@@ -702,6 +702,12 @@ class Parser {
+  /// \p closingBrace is the expected closing brace, which can be either ) or ]
+  /// \p Attr is where to store the parsed attribute
+  bool parseSpecializeAttribute(swift::tok ClosingBrace, SourceLoc AtLoc,
+                                      SourceLoc Loc, SpecializeAttr *&Attr);
```
Nitpick: alignment

slavapestov commented on lib/AST/ArchetypeBuilder.cpp:
```diff
@@ -912,6 +939,63 @@ bool ArchetypeBuilder::addConformanceRequirement(PotentialArchetype *PAT,
+                                            ProtocolDecl *Proto,
+                                            RequirementSource Source,
+                               llvm::SmallPtrSetImpl<ProtocolDecl *> &Visited) {
+  // Add the requirement to the representative.
```
This is just copy and pasted from addConformanceRequirement(). I don't think layouts should be protocols

slavapestov commented on lib/AST/ArchetypeBuilder.cpp:
```diff
@@ -1235,6 +1321,17 @@ bool ArchetypeBuilder::addSameTypeRequirementToConcrete(
       conformances.insert({protocol, *conformance});
     }
+
+    // Make sure the concrete type fulfills the superclass requirement
```
Is this a bug fix? Can you put it in a separate patch?

slavapestov commented on lib/AST/ArchetypeBuilder.cpp:
```diff
@@ -1254,14 +1351,18 @@ bool ArchetypeBuilder::addSameTypeRequirementToConcrete(
     }
   }
 
+  RequirementSource nestedSource(RequirementSource::Redundant, Source.getLoc());
```
This should also be a separate change

slavapestov commented on lib/AST/ArchetypeBuilder.cpp:
```diff
@@ -1435,6 +1536,42 @@ bool ArchetypeBuilder::visitInherited(
+    RequirementSource source(
+        RequirementSource::Explicit,
+        Req.getLayoutConstraintLoc().getSourceRange().Start);
+    if (Req.getLayoutConstraint()->getClassOrBoundGenericClass()) {
```
Why wuld there be a class there?

slavapestov commented on lib/AST/ArchetypeBuilder.cpp:
```diff
@@ -1435,6 +1536,42 @@ bool ArchetypeBuilder::visitInherited(
+      return true;
+    }
+
+#if 1
```
Remove the #if 1

slavapestov commented on lib/AST/ArchetypeBuilder.cpp:
```diff
@@ -1615,6 +1772,12 @@ class ArchetypeBuilder::InferRequirementsWalker : public TypeWalker {
+        } else if (req.getKind() == RequirementKind::Layout) {
+          auto proto = req.getSecondType()->castTo<ProtocolType>();
+          if (Builder.addLayoutRequirement(subjectPA, proto->getDecl(),
+                                                source)) {
```
Nitpick: whitespace

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -2991,9 +2991,11 @@ GenericParamList *ProtocolDecl::createGenericParams(DeclContext *dc) {
   auto protoType = ProtocolType::get(this, ctx);
-  TypeLoc selfInherited[1] = { TypeLoc::withoutLoc(protoType) };
+  //TypeLoc selfInherited[1] = { TypeLoc::withoutLoc(protoType) };
+  auto InheritedTyR = new (ctx) FixedTypeRepr(protoType, this->getLoc());
```
I removed this a while ago, why are we putting a TypeRepr back in here?

slavapestov commented on lib/AST/Type.cpp:
```diff
@@ -2888,7 +2892,33 @@ Optional<ProtocolConformanceRef>
+  if (!conformingReplacementType)
+    return None;
+
+  if (auto nominal =
```
It looks like you're working around a problem in lookupConformance() here? Can you add comments or explain in the review

slavapestov commented on lib/IRGen/GenArchetype.cpp:
```diff
@@ -374,9 +423,12 @@ const TypeInfo *TypeConverter::convertArchetypeType(ArchetypeType *archetype) {
   // representation.
-  if (archetype->requiresClass()) {
+  if (archetype->requiresClass() ||
+      LayoutInfo.isRefCountedObject()) {
```
I think isRefCountedObject() should imply requiresClass() since we also want SIL type lowering to treat such archetypes as loadable

slavapestov commented on lib/IRGen/GenOpaque.cpp:
```diff
@@ -669,6 +669,9 @@ void irgen::emitAssignWithTakeCall(IRGenFunction &IGF,
                             SILType T,
                             Address object) {
+  // TODO: If T is a trivial/POD type, nothing needs to be done.
+  if (T.getObjectType().isTrivial(IGF.getSILModule()))
```
Get this info from IRGen typeinfo, not SIL

slavapestov commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -3524,8 +3524,19 @@ void IRGenSILFunction::emitDebugInfoForAllocStack(AllocStackInst *i,
 }
 
 void IRGenSILFunction::visitAllocStackInst(swift::AllocStackInst *i) {
+  /*
```
Clean this up

slavapestov commented on lib/Parse/ParseGeneric.cpp:
```diff
@@ -274,9 +274,21 @@ ParserStatus Parser::parseGenericWhereClause(
+      // Check if it is a special pseudo-protocol used to express
+      // layout constraints
+      if (isLayoutConstraintType(Protocol.get())) {
+        // TODO: Check if layout constraints are allowed in the
```
This can be checked in ArchetypeBuilder

slavapestov commented on lib/Parse/ParseSIL.cpp:
```diff
@@ -683,6 +689,80 @@ static bool parseSILOptional(bool &Result, SILParser &SP, StringRef Expected) {
+} // end anonymous namespace
+
+/// Remap RequirementReps to Requirements.
+void SILParser::convertRequirements(SILFunction *F,
```
Why does the SIL parser need special code for this?

swiftix replied:
Mainly because we need to perform a proper type lookup for generic type parameters used on the LHS of a requirement, e.g. `T : _Tirival`. The _specialize attribute needs to be type-checked in the scope of the current SIL function, but we do not create FuncDecls for SIL functions. Thus, we need to use parser's generic scopes to resolve those generic type parameters.

slavapestov replied:
We already parse the SIL function types generic signature without such hacks. Can we resolve generic parameters in the specialize attribute the same way?

swiftix replied:
Actually, this hack is based on what we do for SIL function types. When we parse SIL function types they call internally parseTypeIdentifier at some point, which actually does resolve the references to the generic parameters in the current generic scope of the parser. But in case of the `@_specialize` attribute, we parse the references to generic parameters before (!!!) we have parsed the generic signature. Thus, the current generic parser scope does not contain enough information at the time of parsing to resolve such references. Therefore, we need to perform this resolution after we have parsed the generic signature of the SIL function. And when we do it, we essentially mimic in IdentTypeReprLookup and convertRequirements what the SIL parser usually does.

slavapestov replied:
In SIL mode, can we just put the @_specialize attribute after the SIL function signature?

swiftix replied:
In principle we could, though it would look a bit unorthodox ;-) Currently we always print all attributes before the SIL function signature.

swiftix replied:
BTW, in the original version of this code I used a different approach. I created a temporary FuncDecl for the parsed SILFunction and then resolved the requirements in this DeclContext. But that implementation had more code and looked even uglier.

slavapestov commented on lib/Parse/ParseSIL.cpp:
```diff
@@ -683,6 +689,80 @@ static bool parseSILOptional(bool &Result, SILParser &SP, StringRef Expected) {
+namespace {
+  /// A helper class to perform lookup of IdentTypes in the
+  /// current parser scope.
+  class IdentTypeReprLookup : public ASTWalker {
```
This is a hack, it can be avoided by modeling layout requirements better

swiftix replied:
IMHO, it has very little to do with layout requirements. It is required for the lookup of the generic type parameters of the current SIL function. See my comment above.

slavapestov commented on lib/Parse/ParseSIL.cpp:
```diff
@@ -736,19 +816,31 @@ static bool parseDeclSILOptional(bool *isTransparent, bool *isFragile,
-      /// and pushing a new ParsedSpecAttr on the SpecAttrs list. Conformances
-      /// cannot be generated until the function declaration is fully parsed so
-      /// that the function's generic signature can be consulted.
+      //SourceLoc AtLoc = SP.P.consumeToken(tok::identifier);
```
Clean this up

slavapestov commented on lib/SIL/TypeLowering.cpp:
```diff
@@ -1062,11 +1143,16 @@ namespace {
 
-    const TypeLowering *handleTrivial(CanType type) {
+    const TypeLowering *
+    handleTrivial(CanType type,
```
Just add a new handleTrivialAddressOnly() method

slavapestov commented on lib/SILOptimizer/IPO/EagerSpecializer.cpp:
```diff
@@ -127,8 +129,9 @@ static void addReturnValueImpl(SILBasicBlock *RetBB, SILBasicBlock *NewRetBB,
-static void addReturnValue(SILBasicBlock *NewRetBB, SILValue NewRetVal) {
-  auto *RetBB = &*NewRetBB->getParent()->findReturnBB();
+static void addReturnValue(SILBasicBlock *NewRetBB, SILBasicBlock *OldRetBB, SILValue NewRetVal) {
+  //auto *RetBB = &*NewRetBB->getParent()->findReturnBB();
```
Clean this up

slavapestov commented on lib/SILOptimizer/IPO/UsePrespecialized.cpp:
```diff
@@ -88,20 +88,19 @@ bool UsePrespecialized::replaceByPrespecialized(SILFunction &F) {
-    auto SpecType = ReInfo.getSpecializedType();
-    if (!SpecType)
+    // Bail if any generic type parameters are unbound.
+    if (hasUnboundGenericTypes(Subs))
```
Why?

swiftix replied:
We can remove this limitation once we support public partial specializations exposed by other modules. I'll add a TODO comment about this.

slavapestov replied:
What's he actual limitation here?

swiftix replied:
The actual limitation is/was that we only supported pre-specializations for full-specializations so far. There are no principal issues. But I want to wire this up once we have landed all partial specialization related changes. To work properly this code depends on proper mangling, proper visibility, etc.

slavapestov commented on lib/SILOptimizer/Utils/GenericCloner.cpp:
```diff
@@ -73,6 +73,10 @@ void GenericCloner::populateCloned() {
     SILArgument *OrigArg = *I;
     RegularLocation Loc((Decl *)OrigArg->getDecl());
     AllocStackInst *ASI = nullptr;
+    // TODO: Special handling of archetypes with have constraint layouts.
```
They don't have to be passed indirectly. That's sort of the point

swiftix replied:
I only added a comment here, no code changes. And yes, those archetypes will be not be passed indirectly.

IIRC, the logic here is something like:
- the caller used the a generic apply and was passing the parameters indirectly.
- All the code in the non-specialized callee's body is assuming that parameters are passed indirectly.
- then we try to specialize the callee and detect that we can make some parameters concrete.
- when we clone the callee's body during specialization, we rewrite it in such a way that it knows which parameters are not indirect anymore. 
- The way how it is currently modeled is to alloc_stack them in the entry block, which allows for very easy cloning, since all the code which was assuming indirect access doesn't need to be changed. Later on most of this gets promoted into direct access anyways.
- In principle, we could have rewritten indirect access into direct access, but this is a separate issue.

@eeckstein may know more about the reason why we don't rewrite indirect into direct access currently, because he implemented this logic.

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -17,17 +17,94 @@
+    if (!replacement->hasTypeParameter() && !replacement->hasArchetype()) {
+      // It is a concrete type.
+      auto nominal = replacement->getNominalOrBoundGenericNominal();
+      if (nominal) {
```
This looks like it's duplicating your new utility function in ProtocolConformanceRef

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -44,36 +121,54 @@ static unsigned getBoundGenericDepth(Type t) {
+/// For a dependent type of a generic type returns its base
+/// generic type. For a generic type parameter type, returns this
+/// type.
+GenericTypeParamType *getBaseGenericTypeParamType(CanType Ty) {
```
Make this take a Type instead of CanType and canonicalize the result instead of each iteration

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -84,17 +179,114 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
+    return false;
+  }
 
+#if 0
```
Clean this up

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -84,17 +179,114 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
+  // they are very big and are not important for performance.
+  // TODO: Could it be that they are important for string interpolation?
+  // TODO: Introduce a proper way to tell the compiler that certain
+  // functions should not be (auto)-(partially)-specialized.
```
Can you add @_specialize(none) or something for this? :)

swiftix replied:
Yeah. I could do something like this. But a traditional way would be to use `@_semantics` for this purpose, e.g. `@_semantics("optimize.sil.generic.specialization.never")`

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -84,17 +179,114 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
+    return false;
+
+  // We need a generic environment for the partial specialization.
+  if (HasUnboundGenericParams && !Callee->getGenericEnvironment())
```
If the generic callee doesn't have a generic environment, it has no body -- so you won't be specializing it. This check is unnecessary

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -84,17 +179,114 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
+                                     Type Replacement = nullptr) {
+  auto FromCanTy = FromTy->getCanonicalType();
+  auto ToCanTy = ToTy->getCanonicalType();
+  if (isa<SubstitutableType>(FromCanTy)) {
```
Is FromCanTy ever not a SubstitutableType?

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -103,25 +295,957 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
   for (SILParameterInfo PI : SubstitutedType->getParameters()) {
-    if (PI.getSILType().isLoadable(M) &&
+    if (!PI.getSILType().hasTypeParameter() &&
+        PI.getSILType().isLoadable(M) &&
```
Why do you skip type parameters? Class-bound archetypes and such are loadable.

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -103,25 +295,957 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
 }
 
+/// Replace dependent types with their archetypes or concrete types.
+static Type substConcreteTypesForDependentTypes(ModuleDecl &SM,
```
This is just type.subst(SubsMap), and the module is not necessary

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -103,25 +295,957 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
+                                                Type type) {
+  // Cannot use type.subst(SubsMap) here, because it requires a proper
+  // set of conformances to be present in SubsMap, which is not the case.
+
```
... OK, then you should fix this instead of adding hacks :)

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -103,25 +295,957 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
+  for (auto &reqReq : GenSig->getRequirements()) {
+    SubstitutableType *FirstTy =
+        dyn_cast<SubstitutableType>(reqReq.getFirstType()->getCanonicalType());
+    //assert((!FirstTy || SubstMap.getMap().lookup(FirstTy)) &&
```
Why is this commented out?

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -103,25 +295,957 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
+  if (OptimizeGenericSubstitutions) {
+    // Is it an unconstrained generic parameter?
+    if (Conformances.empty()) {
+      if (Replacement->is<ArchetypeType>() ||
```
Why are you getting both archetypes and interface types in this function?

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -103,25 +295,957 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
+        CallerGenericEnv->mapTypeOutOfContext(CallerArchetype)
+            ->getCanonicalType();
+    assert(CallerGenericParam);
+    assert(CallerGenericParam->is<GenericTypeParamType>());
```
How do you know it's not a member type?

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -103,25 +295,957 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
+
+namespace {
+template<typename SubstFn>
+struct SubstDependentSILType
```
Isn't this just copy and pasted from SILFunctionType.cpp? What's different about this version?

slavapestov commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -950,6 +950,25 @@ void ConstraintSystem::openGeneric(
+      auto proto = req.getSecondType()->castTo<ProtocolType>();
+      auto protoDecl = proto->getDecl();
+
+      // Determine whether this is the protocol 'Self' constraint we should
```
This check does not apply here

slavapestov commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -324,6 +324,33 @@ void TypeChecker::checkGenericParamList(ArchetypeBuilder *builder,
 
+    case RequirementReprKind::LayoutConstraint: {
+      // Validate the types.
+      if (validateType(req.getSubjectLoc(), lookupDC, options, resolver)) {
```
This is also copy and pasted...

slavapestov commented on lib/Serialization/Serialization.h:
```diff
@@ -476,6 +472,10 @@ class Serializer {
+
+  /// Writes a set of generic requirements.
+  void writeGenericRequirements(ArrayRef<Requirement> requirements,
+                                const std::array<unsigned, 256> &abbrCodes);
```
This change is unnecessary

slavapestov commented:
Roman, since this PR is too large to review do you mind splitting it up into smaller pieces?

- One patch introducing the new specialized syntax, but makes it a no-op for now.

- Patch introducing any new substitution-related utility methods in common code.

- Patch introducing new partial specialization machinery.

- Patch introducing new layout constraints.

Even if it takes a bit more work to split things up and get each piece to work on its own, it will make it much easier to review and test each patch.

slavapestov commented:
Tip:

    git reset --soft HEAD^
    git reset HEAD
    git add -p
    ... record first patch
    git commit -m "first patch"
    git add -p
    ... record second patch
    git commit -m "second patch"
    etc

swiftix commented:
@slavapestov Yes, I agree that it could be split, but it would take some time.

Thinking in this direction, this PR is based on another smaller PR for the partial specialization #5463. That one has interesting code changes mostly in Generics.cpp. More over, it is self-contained. Could you review that one in the meantime, while I work on splitting this one into smaller chunks? #5463  would probably cover 1/4 or 1/3 of the current PR.

slavapestov commented:
I agree it will take time but it will improve the quality of the final code since it still appears we need to iterate on several aspects of it.

swiftix commented:
@slavapestov Yes, we are in agreement. I haven not meant that I won't do it because it would take time. I just meant that I'd need time for it.

swiftix committed:
Many changes related to the representation of layout constraints + code clean-ups.

- Introduce LayoutConstaint
- Do not model layout requirements as conformances to special protocols. Use LayoutConstraint instead to represent them.
- Address review comments (remove a lot of dead code, cleanups, etc)

slavapestov commented:
Thank you! This is a great change overall by the way, we just need to sort out the rough edges.

swiftix commented:
@slavapestov I just pushed the current non-chunked version. It is rebased and address many minor issues you noticed. The most important change is that it does not use types to represent layout constraints and uses a dedicated LayoutConstraint datatype to represent them.

slavapestov commented:
Ok, I'll take a look, thanks. The main reason I want it chunked is to understand what utility operations you're adding for substitutions. Some might be unnecessary, others might be working around bugs we can fix at the root cause and others suggest the API of SubstitutionMap is deficient.

slavapestov commented:
For example, adding a generic signature to SILFunctionType::substGenericArgs makes sense to me since we need it for lowering replacement types but if it's a separate patch then Joe could also review it, etc.

swiftix committed:
Fix the test

swiftix commented:
@swift-ci Please smoke test

swiftix commented:
@swift-ci Please smoke test

swiftix committed:
Fix a bug in SILPrinter

slavapestov commented on include/swift/AST/Requirement.h:
```diff
@@ -47,7 +47,10 @@ enum class RequirementKind : unsigned {
 class Requirement {
   llvm::PointerIntPair<Type, 3, RequirementKind> FirstTypeAndKind;
-  Type SecondType;
+  union {
```
Could this be a PointerUnion for type safety?

swiftix replied:
In principle, it could. At least right now, because we have only two possibilities. Should we add more in the future, we may run into problems. Besides, all the getters are behind assertions currently, which check the kind of the requirement before accessing the elements of this union.

slavapestov commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -1318,11 +1321,13 @@ void PrintAST::printSingleDepthOfGenericSignature(
+          if (!(first->is<ArchetypeType>() || first->isTypeParameter()) &&
+              !(second->is<ArchetypeType>() || second->isTypeParameter()))
+            continue;
+        }
```
Maybe we should still print the layout constraint here if second is null

slavapestov commented on lib/AST/ArchetypeBuilder.cpp:
```diff
@@ -2077,26 +2000,31 @@ void ArchetypeBuilder::dump(llvm::raw_ostream &out) {
   enumerateRequirements([&](RequirementKind kind,
                             PotentialArchetype *archetype,
-                            llvm::PointerUnion<Type, PotentialArchetype *> type,
+                            llvm::PointerUnion3<Type, PotentialArchetype *, LayoutConstraint> constraint,
```
A typedef for this PointerUnion3 would be nice, call it RequirementRHS maybe

swiftix replied:
Sure. Good idea.

slavapestov commented on lib/AST/LayoutConstraint.cpp:
```diff
@@ -58,44 +39,23 @@ LayoutConstraintInfo getLayoutConstraintInfo(StringRef ID) {
+StringRef LayoutConstraintInfo::getName() const {
+  switch (getKind()) {
+  case LayoutConstraintKind::UnknownLayout:
+    return "_UnknwonLayout";
```
Typo: Unknwon

You might want to try using a .def file to define layout constraint kinds, like we do for types and decls etc.

swiftix replied:
Hmm. This is an interesting idea. I'll see if I can do something about it.

slavapestov commented on lib/AST/Type.cpp:
```diff
@@ -1594,6 +1597,13 @@ bool TypeBase::isSpelledLike(Type other) {
 
+LayoutConstraint TypeBase::getLayoutConstraint() {
+  auto CanTy = getCanonicalType();
+  if (!CanTy)
```
getCanonicalType() never returns nullptr

swiftix replied:
Done.

slavapestov commented on lib/AST/Type.cpp:
```diff
@@ -673,8 +673,11 @@ ClassDecl *CanType::getClassBoundImpl(CanType type) {
 
-LayoutConstraintInfo CanType::getLayoutConstraintInfo() const {
-  return swift::getLayoutConstraintInfo(*this);
+LayoutConstraint CanType::getLayoutConstraint() const {
```
There's no need to canonicalize the type first -- just do if (auto archetypeTy = getAs<ArchetypeType>(this)) { ... }

swiftix replied:
Done.

slavapestov commented on lib/AST/Type.cpp:
```diff
@@ -2941,6 +2962,8 @@ LookUpConformanceInSubstitutionMap::operator()(CanType dependentType,
     return None;
 
+  // TODO: Describe what this code is doing. It works around the issues
+  // in the conformance lookup, when a dependent type cannot be found...
```
It sounds like either you're not adding the right conformances when building a SubstitutionMap, or SubstitutionMap::lookupConformance() is not returning the right thing.

I'm planning on re-doing SubstitutionMap construction so that it is driven by callbacks rather than calls to addSubstitution()/addConformance() on the map. This will ensure the correct entries are always added.

slavapestov commented on lib/Parse/ParseType.cpp:
```diff
@@ -59,6 +59,81 @@ TypeRepr *Parser::applyAttributeToType(TypeRepr *ty,
+  int alignment = 0;
+
+  auto ParseTrivialLayoutConstraintBody = [&] () -> bool {
+    // Parse the size and alginment.
```
Typo: alignment

swiftix replied:
Fixed.

slavapestov commented on lib/Parse/ParseType.cpp:
```diff
@@ -59,6 +59,81 @@ TypeRepr *Parser::applyAttributeToType(TypeRepr *ty,
+
+  // Otherwise it is a trivial layout constraint with
+  // provided size and alginment.
+  return LayoutConstraintInfo(layoutConstraint.getKind(), size, alignment);
```
Some tests for various invalid cases in test/attr/attr_specialize.swift would be great!

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -91,14 +91,16 @@ class QueryConformance {
         SmallVector<ProtocolConformance *, 2> conformances;
```
When does the lookupConformance() call above fail?

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -124,13 +126,13 @@ static unsigned getBoundGenericDepth(Type t) {
-  assert(isa<GenericTypeParamType>(BaseTy));
-  return cast<GenericTypeParamType>(BaseTy);
+  assert(BaseTy->is<GenericTypeParamType>());
+  return BaseTy->getAs<GenericTypeParamType>();
```
castTo() is effectively assert(is)+getAs

slavapestov commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -1494,28 +1494,33 @@ checkSpecializeAttrRequirements(SpecializeAttr *attr,
-            .highlight(req.getSourceRange());
-        continue;
-      }
+        if (!constraint || constraint->is<ErrorType>() ||
```
hasError() maybe

slavapestov commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -1494,28 +1494,33 @@ checkSpecializeAttrRequirements(SpecializeAttr *attr,
+              .highlight(req.getSourceRange());
+        }
+        // TODO: Check that it is one of the compiler known protocols used for
+        // pre-specializations.
```
Is this comment still needed?

swiftix replied:
Of course this comment is outdated. Removed.

slavapestov commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -1732,7 +1762,8 @@ void AttributeChecker::visitSpecializeAttr(SpecializeAttr *attr) {
+  checkSpecializeAttrRequirements(attr, FD, resolvedRequirements,
+                                  constrainedGenericParams, TC);
 
   // Store converted requirements in the attribute so that they are
```
This code goes through a lot of trouble to build the convertedRequirements. However much of the logic could go away if we just used the generic signature from the Builder instead. It seems we're only using the ArchetypeBuilder for diagnostics, and then duplicating a lot of work to canonicalize the requirements.

swiftix replied:
Yeah. I'm not quite sure about. We need to figure out if it is possible. But some of the checks can be moved into ArchetypeBuilder I guess.

slavapestov commented:
The new representation of layout constraints looks much cleaner, thanks! I left some review comments suggesting some minor improvements.

Two things I want us to figure out next is whether the specialize attribute should store the specialized generic signature instead of the list of requirements, and cleaning up some of the SIL substitution stuff.

swiftix committed:
Implement mangling of specializations with layout constraints.

swiftix committed:
wip

swiftix committed:
Add a helper method for debugging mangling and demangling

swiftix committed:
Adjust after rebasing

swiftix committed:
Fixes after re-basing

swiftix commented:
Pushed latest state after re-basing.

Will split into smaller chunks now.

slavapestov commented:
@swiftix Time to close this?

swiftix commented:
Obsolete now. Closing.
